src/module_3/push_exploration.md

```diff
@@ -0,0 +1,5364 @@
+
+
+
+
+```python
+# Validate the model
+y_val_pred = model.predict(X_val)
+print("Classification Report:")
+print(classification_report(y_val, y_val_pred))
+
+print("Confusion Matrix:")
+cm = confusion_matrix(y_val, y_val_pred)
+print(cm)
+```
+
+    Classification Report:
+                  precision    recall  f1-score   support
+    
+               0       0.99      0.78      0.87    213182
+               1       0.04      0.61      0.08      3213
+    
+        accuracy                           0.78    216395
+       macro avg       0.52      0.70      0.48    216395
+    weighted avg       0.98      0.78      0.86    216395
+    
+    Confusion Matrix:
+    [[166533  46649]
+     [  1244   1969]]
+
+
+
+```python
+# Calculate probabilities for the positive class
+y_val_probs = model.predict_proba(X_val)[:, 1]
+
+# ROC Curve
+fpr_val, tpr_val, _ = roc_curve(y_val, y_val_probs)
+roc_auc_val = auc(fpr_val, tpr_val)  # Calculate AUC
+
+plt.figure(figsize=(10,5))
+
+plt.subplot(1, 2, 1)
+plt.plot(fpr_val, tpr_val, label=f'ROC (AUC = {roc_auc_val:.2f})')
+plt.plot([0, 1], [0, 1], color='navy', linestyle='--')
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('False Positive Rate')
+plt.ylabel('True Positive Rate')
+plt.title('ROC')
+plt.legend(loc="lower right")
+
+
+# Precision-Recall Curve
+precision_val, recall_val, _ = precision_recall_curve(y_val, y_val_probs)
+
+plt.subplot(1, 2, 2)
+plt.plot(recall_val, precision_val)
+plt.xlim([0.0, 1.0])
+plt.ylim([0.0, 1.05])
+plt.xlabel('Recall')
+plt.ylabel('Precision')
+plt.title('Precision-Recall Curve')
+
+plt.tight_layout()
+plt.show()
+```
+
+
+    
+![png](push_exploration_files/push_exploration_44_0.png)
+    
+
+
+Better results with this approach, where we lose precision but gain sensitivity for purchases.
```

Comment:
How are these better results? I see the exact same AUC. The PR curve has slightly? changed but can't see how is it better now (maybe you can get the auc PR)? So no, I wouldn't claim the results are better (which confirms my theory that class unbalance is normally BS)